### ISSUE TITLE
aggregation index can become larger than 100

### ISSUE THREAD

**Issue by dschlaep** [r-spatialecology/landscapemetrics#93]
Problem: The aggregation index can become larger than 100.
Expected range: 0-100 (as per documentation `?lsm_c_ai` and `?lsm_l_ai`)

Version from March 18, 2019: `devtools::install_github("r-spatialecology/landscapemetrics")`

Reproducible example:
- Create raster with one class --> no problem, as expected
```
r <- raster::raster(nrow = 10, ncol = 10)
r[] <- 1

landscapemetrics::lsm_c_ai(r)$value # result: 100
landscapemetrics::lsm_l_ai(r)$value # result: 100
```

- Add at least one empty cell to raster --> result is larger than maximum expected value
```
r[1, 1] <- NA

landscapemetrics::lsm_c_ai(r)$value # result: 100.5618
landscapemetrics::lsm_l_ai(r)$value # result: 100.5618
```

My hunch: The calls by the function `landscapemetrics:::lsm_c_ai_calc`
to `rcpp_get_coocurrence_matrix_diag` and to `rcpp_get_composition_vector`
treat NAs differently which leads to inconsistencies in the ratio
`like_adjacencies/max_adj`

- raster with once class --> no problem
```
r <- raster::raster(nrow = 10, ncol = 10)
r[] <- 1

landscape <- raster::as.matrix(r)
like_adjacencies <- landscapemetrics:::rcpp_get_coocurrence_matrix_diag(
  landscape, directions = as.matrix(4))/2 # 180
cells_class <- landscapemetrics:::rcpp_get_composition_vector(landscape) # 100
n <- trunc(sqrt(cells_class)) # 10
m <- cells_class - n^2 # 0
max_adj <- 2 * n * (n - 1) # (m == 0) -> 180
```

- Add at least one empty cell --> result is larger than maximum expected value
```
r[1, 1] <- NA
landscape <- raster::as.matrix(r)
like_adjacencies <- landscapemetrics:::rcpp_get_coocurrence_matrix_diag(
  landscape, directions = as.matrix(4))/2 # 179 instead of 178?
cells_class <- landscapemetrics:::rcpp_get_composition_vector(landscape) # 99
n <- trunc(sqrt(cells_class)) # 9
m <- cells_class - n^2 # 18
max_adj <- 2 * n * (n - 1) + 2 * m - 2 # (m > n) -> 178 instead of 179?
```

Thanks for a great package!

**Comment by mhesselbarth** [comment]
Hey Daniel, 

thanks a lot for your very detailed issue. 

We will look into this and come back to you as soon as possible.

**Comment by mhesselbarth** [comment]
Okay....I think I fixed it (113ea720). Not sure if the solution is smart, but it's working 🤷‍♂️  . I also cross-checked with FRAGSTATS quickly, and results seem to be identical. 

Just install the latest development version using `devtools::install_github("r-spatialecology/landscapemetrics"` - will also be part of the next CRAN update.

```
r <- raster::raster(nrow = 10, ncol = 10)

r[] <- 1
landscapemetrics::lsm_c_ai(r)$value
[1] 100

r[1, 1] <- NA
landscapemetrics::lsm_c_ai(r)$value
[1] 100

r[2, 5] <- NA
landscapemetrics::lsm_c_ai(r)$value
[1] 98.86364
```

@marcosci @Nowosad: Please have a look, maybe it's smarter/worth to somehow modify the Rcpp stuff?

**Comment by dschlaep** [comment]
Great! Thank you for addressing this so quickly!

**Comment by bitbacchus** [comment]
The bug was actually in the C++ code. I fixed it (8968bb6) hopefully.

**Comment by mhesselbarth** [comment]
Well done 👍 

Feel free to re-open at any time.
